dbt_subprojects/dex/macros/models/_project/uniswap_compatible_liquidity.sql

```diff
@@ -0,0 +1,141 @@
+{% macro uniswap_compatible_v4_liquidity(
+    blockchain = null
+    , project = 'uniswap'
+    , version = '4'
+    , PoolManager_evt_ModifyLiquidity = null
+    , PoolManager_evt_Swap = null
+    , PoolManager_evt_Initialize = null
+    , pair_column_name = 'id'
+    )
+%}
+
+WITH filtered_modify_liquidity AS (
+    SELECT 
+        evt_block_time,
+        evt_block_number,
+        id,
+        evt_tx_hash,
+        evt_index,
+        salt,
+        tickLower,
+        tickUpper,
+        liquidityDelta
+    FROM {{ PoolManager_evt_ModifyLiquidity }}
+    {%- if is_incremental() %}
+    WHERE {{ incremental_predicate('evt_block_time') }}
+    {%- endif %}
+),
+
+filtered_swaps AS (
+    SELECT 
+        evt_block_time,
+        id,
+        sqrtPriceX96
+    FROM {{ PoolManager_evt_Swap }}
+    {%- if is_incremental() %}
+    WHERE {{ incremental_predicate('evt_block_time') }}
+    {%- endif %}
+),
+
+filtered_initialize AS (
+    SELECT 
+        evt_block_time,
+        {{pair_column_name}} as id,
+        currency0,
+        currency1,
+        sqrtPriceX96
+    FROM {{ PoolManager_evt_Initialize }}
+),
+
+get_recent_sqrtPriceX96 AS (
+    SELECT *
+    FROM (
+        SELECT 
+            ml.*
+            ,i.currency0 as token0
+            ,i.currency1 as token1
+            ,COALESCE(s.evt_block_time, i.evt_block_time) as most_recent_time
+            ,COALESCE(s.sqrtPriceX96, i.sqrtPriceX96) AS sqrtPriceX96
+            ,ROW_NUMBER() OVER (
+                PARTITION BY ml.id, ml.evt_block_time
+                ORDER BY 
+                    CASE WHEN s.sqrtPriceX96 IS NOT NULL THEN s.evt_block_time ELSE i.evt_block_time END DESC
+            ) AS rn
+        FROM filtered_modify_liquidity ml
+        LEFT JOIN filtered_swaps s 
+            ON ml.evt_block_time > s.evt_block_time AND ml.id = s.id
+        LEFT JOIN filtered_initialize i 
+            ON ml.evt_block_time >= i.evt_block_time AND i.id = ml.id
```

Comment:
Though I am not fully understand the joining condition,
`ml.evt_block_time > s.evt_block_time` means, liquidity must set after the swap_evt?